### PR TITLE
Add lexer for the Savi language.

### PR DESCRIPTION
**Pull request by jemc**
This PR adds a lexer for the [Savi programming language](https://github.com/savi-lang/savi).

### CODE REVIEW

File: pygments/lexers/savi.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    (r'//.*?$', Comment.Singleline),
+
+    # Doc Comment
+    (r'::.*?$', Comment.Singleline),
+
+    # Capability Operator
+    (r'(\')(\w+)(?=[^\'])', bygroups(Operator, Name)),
+
+    # Double-Quote String
+    (r'\w?"', String.Double, "string.double"),
```
Comment (Anteru) [inline pygments/lexers/savi.py:50]: Please use `+`  and not `?` so multiple tokens get matched together.

File: pygments/lexers/savi.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  For `Savi <https://github.com/savi-lang/savi>`_ source code.
+  """
+
+  name = 'Savi'
+  aliases = ['savi']
+  filenames = ['*.savi']
+
+  root_rules = [
+    # Line Comment
+    (r'//.*?$', Comment.Singleline),
```
Comment (Anteru) [inline pygments/lexers/savi.py:41]: `Comment.Single`

File: pygments/lexers/savi.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    (r'(\])(\!)', bygroups(Punctuation, Generic.Deleted), "#pop"),
+    (r'\]', Punctuation, "#pop"),
+
+    # Expression Separators
+    (r'(,|;|:)', Punctuation),
+
+    # Other "Punctuation"
+    (r'(@|\.)', Punctuation),
+
+    # Piping Operators
```
Comment (Jean-Abou-Samra) [inline pygments/lexers/savi.py:103]: Same here, you can merge all operators in a single rule.

**Review by Jean-Abou-Samra — COMMENTED** [review]
(Note that I am not a maintainer, just someone lurking around and hoping that a preliminary review will help. You could also review my own PR ;-)

(Update: "preliminary review" was written before I reloaded the page and saw @anteru had already left some comments.)

File: pygments/lexers/savi.py (hunk 1 above)
Comment (jemc) [reply pygments/lexers/savi.py:50]: This is actually a single optional `\w` character accepted here, not multiple - note that this is immediately before the string, not inside the string.

The optional `\w` here enables things like `b"example"` for a byte string, or `r"example"` for a regular expression string (similar to the one seen here in the python source code).

Comment (Anteru) [reply pygments/lexers/savi.py:50]: Ah, my bad. I'm looking at this here in the output:
```
'"'           Literal.String.Double
's'           Literal.String.Double
'p'           Literal.String.Double
'e'           Literal.String.Double
'c'           Literal.String.Double
'"'           Literal.String.Double
```

And those `s`, `p`, `e`, `c` should become one token. I just scanned for `String.Double` and found the wrong rule :)

Comment (jemc) [reply pygments/lexers/savi.py]: Yes, I've fixed that issue in my latest commit.

File: pygments/lexers/savi.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+      (r'(\=)', Operator),
+
+      # Other Operators
+      (r'(\!|\<\<|\<|\&|\|)', Operator),
+
+      # Identifiers
+      (r'\b\w+\b', Name),
+
+      # Whitespace
+      (r'[ \t\r]+', Text),
```
Comment (amitkummer) [inline pygments/lexers/savi.py:125]: Same.

File: pygments/lexers/savi.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+      (r'\w?"', String.Double, "string.double"),
+
+      # Single-Char String
+      (r"'", String.Char, "string.char"),
+
+      # Class (or other type)
+      (r'([_A-Z]\w*)', Name.Class),
+
+      # Declare
+      (r'^(\s*)(:)(\w+)',
```
Comment (amitkummer) [inline pygments/lexers/savi.py:62]: There is a `Whitespace` token for this :)

File: pygments/lexers/savi.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+      # Whitespace
+      (r'[ \t\r]+', Text),
+    ],
+
+    # Declare (nested rules)
+    "decl": [
+      (r'\b[a-z_]\w*\b(?!\!)', Keyword.Declaration),
+      (r':', Punctuation, "#pop"),
+      (r'\n', Text, "#pop"),
```
Comment (amitkummer) [inline pygments/lexers/savi.py:132]: Same.

File: pygments/lexers/savi.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+# from the Savi repository flows forward into the pygments repository.
+#
+# If you're changing this file in the Savi repository, please ensure that
+# any changes you make are also reflected in the other pygments-like lexers
+# (rouge, vscode, etc) so that all of the lexers can be kept cleanly in sync.
+
+class SaviLexer(RegexLexer):
+  """
+  For `Savi <https://github.com/savi-lang/savi>`_ source code.
+  """
```
Comment (Anteru) [inline pygments/lexers/savi.py:35]: This needs a `.. versionadded: 2.10` marker.

**Review by Anteru — COMMENTED** [review]
This lexer matches colons as Tag or Punctuation. Is that by design? It seems quite confusing.

**Comment by jemc** [comment]
> This lexer matches colons as Tag or Punctuation. Is that by design? It seems quite confusing.

Yes, it's by design. When a colon and immediately adjacent identifier are at the start of a line, together they form a tag that begins a declaration. Elsewhere the colon appears outside that one particular context, it is punctuation.

File: pygments/lexers/savi.py (hunk 7 above)
Comment (jemc) [reply pygments/lexers/savi.py:37]: Added in 63c1f634eaf3785f4584948b903ccf4adc291ea3

**Review by Anteru — APPROVED** [review]

**Comment by Anteru** [comment]
Merged, thanks a lot!

**Comment by jemc** [comment]
@Anteru - Thanks for merging this!

Is there another release coming up soon that could make this lexer available to downstream tools?

**Comment by Anteru** [comment]
There is, I'm just swamped right now :(

**Pull request merged**